internal/grpc/services/usershareprovider/usershareprovider.go

```diff
@@ -497,40 +506,181 @@ func (s *service) GetReceivedShare(ctx context.Context, req *collaboration.GetRe
+	updatedShare, err := s.sm.UpdateReceivedShare(ctx, req.Share, req.UpdateMask, &uid)
 	if err != nil {
 		return &collaboration.UpdateReceivedShareResponse{
 			Status: status.NewInternal(ctx, "error updating received share"),
 		}, nil
 	}
 
-	res := &collaboration.UpdateReceivedShareResponse{
+	return &collaboration.UpdateReceivedShareResponse{
 		Status: status.NewOK(ctx),
-		Share:  share,
+		Share:  updatedShare,
+	}, nil
+}
+
+// GetAvailableMountpoint returns a new or existing mountpoint
+func GetAvailableMountpoint(ctx context.Context, gwc gateway.GatewayAPIClient, id *provider.ResourceId, name string) (string, error) {
+	listReceivedSharesRes, err := gwc.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{})
+	if err != nil {
+		return "", errtypes.InternalError("grpc list received shares request failed")
 	}
-	return res, nil
+
+	if err := errtypes.NewErrtypeFromStatus(listReceivedSharesRes.GetStatus()); err != nil {
+		return "", err
+	}
+
+	// we need to sort the received shares by mount point in order to make things easier to evaluate.
+	base := filepath.Clean(name)
+	mount := base
+	existingMountpoint := ""
+	mountedShares := make([]string, 0, len(listReceivedSharesRes.GetShares()))
+	var pathExists bool
+
+	for _, s := range listReceivedSharesRes.GetShares() {
+		resourceIDEqual := utils.ResourceIDEqual(s.GetShare().GetResourceId(), id)
+
+		if resourceIDEqual && s.State == collaboration.ShareState_SHARE_STATE_ACCEPTED {
+			// a share to the resource already exists and is mounted, remembers the mount point
+			_, err := utils.GetResourceByID(ctx, s.GetShare().GetResourceId(), gwc)
+			if err == nil {
+				existingMountpoint = s.GetMountPoint().GetPath()
+			}
+		}
+
+		if s.State == collaboration.ShareState_SHARE_STATE_ACCEPTED {
+			// collect all accepted mount points
+			mountedShares = append(mountedShares, s.GetMountPoint().GetPath())
+			if s.GetMountPoint().GetPath() == mount {
+				// does the shared resource still exist?
+				_, err := utils.GetResourceByID(ctx, s.GetShare().GetResourceId(), gwc)
+				if err == nil {
+					pathExists = true
+				}
+				// TODO we could delete shares here if the stat returns code NOT FOUND ... but listening for file deletes would be better
+			}
+		}
+	}
+
+	if existingMountpoint != "" {
+		// we want to reuse the same mountpoint for all unmounted shares to the same resource
+		return existingMountpoint, nil
+	}
+
+	// If the mount point really already exists, we need to insert a number into the filename
+	if pathExists {
+		// now we have a list of shares, we want to iterate over all of them and check for name collisions agents a mount points list
+		for i := 1; i <= len(mountedShares)+1; i++ {
+			ext := filepath.Ext(base)
+			name := strings.TrimSuffix(base, ext)
+			// be smart about .tar.(gz|bz) files
+			if strings.HasSuffix(name, ".tar") {
+				name = strings.TrimSuffix(name, ".tar")
+				ext = ".tar" + ext
+			}
```

Comment:
I don't quite understand this. What is so special with `.tar` files? Could you brievly explain?

Reply:
resources with a suffix needs the counter before the suffix, but true, it is not tar related, same is for zip, .psd, ...

https://github.com/cs3org/reva/blob/40abb344337b95e2af8f3384fbd73c87f4ffdb9a/internal/http/services/owncloud/ocs/handlers/apps/sharing/shares/pending.go#L173

needs a followup

Reply:
How can we make sure this isn't forgotten? This will be a nasty bug (should it ever occur).

Reply:
i start working on it right after that one is merged... but to be honest, this sits here since ages which makes me unhappy because your 100% right!

Reply:
I guess this is because it has 2 file extension? `tar.<dot-something>` ?

Still, I don't think we should do this. I know it has been this way in ocs but is wrong in so many ways ..